### ISSUE TITLE
BadAlloc in eglCreatePlatformWindowSurface if run after Vulkan

### ISSUE THREAD

**Issue by niklaskorz** [gfx-rs/wgpu#1153]
Short info header:
- GFX version: gfx-rs/gfx#3591
- OS: Fire OS 6 (Android 7.1 / API Level 25)
- GPU: PowerVR GE8300 ([Fire TV Stick 4K - 1st Gen (2018)](https://developer.amazon.com/de/docs/fire-tv/device-specifications-fire-tv-stick.html?v=ftvstick4k))

The GL backend fails to create a window surface if run after the Vulkan backend (which isn't supported on the tested device).
While technically the [GPU does support Vulkan 1.0 and 1.1](https://www.imaginationtech.com/powervr-gpu/ge8300/), Fire OS 6 does not. Android itself comes with [Vulkan on API Level 24 and above](https://developer.android.com/ndk/guides/graphics/getting-started), so Amazon probably just didn't bother to include the relevant drivers for the PowerVR GPU.

Creation of the EGL surface succeeds if the Vulkan backend is not run before.

An excerpt from the log:

```
gfx_backend_vulkan: Unable to find extension: VK_EXT_debug_utils
gfx_backend_vulkan: Unable to find extension: VK_KHR_get_physical_device_properties2
gfx_backend_vulkan: Unable to find layer: VK_LAYER_KHRONOS_validation
gfx_backend_gl::window::egl: Client extensions: ""
gfx_backend_gl::window::egl: Using default platform
gfx_backend_gl::window::egl: Display vendor "Android", version (1, 4), extensions: "EGL_KHR_get_all_proc_addresses EGL_ANDROID_presentation_time EGL_KHR_swap_buffers_with_damage EGL_ANDROID_create_native_client_buffer EGL_ANDROID_front_buffer_auto_refresh EGL_KHR_image EGL_KHR_image_base EGL_KHR_gl_texture_2D_image EGL_KHR_gl_texture_cubemap_image EGL_KHR_gl_renderbuffer_image EGL_KHR_fence_sync EGL_KHR_create_context EGL_KHR_surfaceless_context EGL_EXT_create_context_robustness EGL_ANDROID_image_native_buffer EGL_KHR_wait_sync EGL_ANDROID_recordable EGL_KHR_partial_update EGL_EXT_buffer_age EGL_KHR_mutable_render_buffer EGL_EXT_yuv_surface "
gfx_backend_gl::window::egl: Error in create_platform_window_surface: BadAlloc
```

**Comment by Gordon-F** [comment]
@niklaskorz  [create_platform_window_surface](https://developer.android.google.cn/reference/android/opengl/EGL15#eglCreatePlatformWindowSurface(android.opengl.EGLDisplay,%20android.opengl.EGLConfig,%20java.nio.Buffer,%20long[],%20int)) added in API level 29 (Android 10). Looks like your device supports only EGL 1.4.

Are you running your app with the gfx master branch or from fork https://github.com/gfx-rs/gfx/pull/3591?

**Comment by nmpribeiro** [comment]
I am seeing a somewhat similar bug

```
I/rust-app: gfx_backend_vulkan: Unable to find extension: VK_EXT_debug_utils
    gfx_backend_vulkan: Unable to find extension: VK_KHR_get_physical_device_properties2
    gfx_backend_vulkan: Unable to find extension: VK_EXT_display_surface_counter
    gfx_backend_vulkan: Unable to find extension: VK_KHR_display
I/vulkan: Loaded layer VK_LAYER_KHRONOS_validation
I/RustStdoutStderr: thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', src/app.rs:56:14
    stack backtrace:
    note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
I/rust-app: wgpu_core::hub: Dropping Global
I/RustStdoutStderr: attempt to unwind out of `rust` with err: Any { .. }
A/libc: Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 1911 (ro.rust_app), pid 1864 (ro.rust_app)
```

At `src/app.rs:56:14` I have the following:

```
        let adapter = instance
            .request_adapter(&wgpu::RequestAdapterOptions {
                power_preference: wgpu::PowerPreference::default(),
                compatible_surface: Some(&surface),
            })
            .await
            .unwrap();
```

This is following a wgpu triangle demo somewhere. Do you have any ideas about it?

**Comment by nmpribeiro** [comment]
I mean, I can detect that the adapter is None, but shouldn't wgpu do it kinda automatically and drop to openGL or something?

**Comment by Gordon-F** [comment]
@nmpribeiro Are you running on wgpu 0.9 (gfx-hal) or master (wgpu-hal) ?

Make sure you requesting all backends:
https://github.com/gfx-rs/wgpu/blob/b8d5c9d9acb6e3c67a30c219257bb46878db0bbb/wgpu/examples/hello-triangle/main.rs#L10

**Comment by nmpribeiro** [comment]
Hi @Gordon-F !

I am using wgpu:
```
[dependencies]
wgpu = "0.9.0"
...
```

I don't see `wgpu::Backends::all()`. There is `wgpu::BackendBit::all()` and `wgpu::Backend::{GL, BrowserWebGpu, Dx11, Dx12, Empty, Metal, Vulkan}` but no `all()` there.

Also, I have tried `let instance = wgpu::Instance::new(wgpu::BackendBit::all());` and had the following runtime error:

```
I/rust-app: wgpu_core::instance: Adapter GL AdapterInfo { name: "Android Emulator OpenGL ES Translator (AMD Radeon Pro 5500M OpenGL Engine)", vendor: 0, device: 0, device_type: DiscreteGpu }
W/rust-app: wgpu_core::instance: The underlying API or device in use does not support enough features to be a fully compliant implementation of WebGPU. A subset of the features can still be used. If you are running this program on native and not in a browser and wish to limit the features you use to the supported subset, call Adapter::downlevel_properties or Device::downlevel_properties to get a listing of the features the current platform supports.
    wgpu_core::instance: Missing internal features: ROBUST_BUFFER_ACCESS | IMAGE_CUBE_ARRAY | INDEPENDENT_BLENDING | SAMPLE_RATE_SHADING | SAMPLER_ANISOTROPY | FRAGMENT_STORES_AND_ATOMICS
I/rust-app: wgpu_core::command::allocator: Starting on (internal) thread ThreadId(2)
    gfx_backend_gl::device: 	Created frame buffer 1
W/rust-app: wgpu_core::device::alloc: max_memory_allocation_count is not set by gfx-rs backend
I/RustStdoutStderr: thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', /Users/nmpribeiro/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-core-0.9.2/src/hub.rs:878:29
    stack backtrace:
    note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
I/rust-app: wgpu_core::hub: Dropping Global
W/libEGL: eglTerminate() called w/ 1 objects remaining
I/RustStdoutStderr: attempt to unwind out of `rust` with err: Any { .. }
A/libc: Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 4177 (ro.rust_app), pid 4114 (ro.rust_app)
```

Using `let instance = wgpu::Instance::new(wgpu::BackendBit::GL);` works, but once I lose focus of the app (by hitting android square button) I get:

```
E/EGL_emulation: tid 4379: swapBuffers(681): error 0x300d (EGL_BAD_SURFACE)
I/RustStdoutStderr: thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: BadSurface', /path/to/my/.cargo/registry/src/github.com-1ecc6299db9ec823/gfx-backend-gl-0.9.0/src/window/egl.rs:667:55
```

Should I change to `wgpu-hal`?

**Comment by nmpribeiro** [comment]
Btw, I was using `let instance = wgpu::Instance::new(wgpu::BackendBit::PRIMARY);`

**Comment by kvark** [comment]
Should be fixed by #1656. The `eglCreatePlatformWindowSurface` is now called at `configure()` time.
